File: app/briefing.py
"""
Builds the Telegram messages (Markdown). All dynamic text is run through
sanitize_markdown so it can't break Telegram's parser.
"""

from __future__ import annotations

from datetime import date

from app.telegram import sanitize_markdown as _s

SEV_EMOJI = {"CRITICAL": "🔴", "HIGH": "🟠", "MEDIUM": "🟡", "LOW": "⚪"}

PIPELINE_LINE = "_Pipeline: Gemini 2.5 Flash + Flash-Lite · Groq Llama 4 Scout_"


def _today() -> str:
    return date.today().strftime("%A, %-d %B")


def _footer(app_url: str) -> str:
    return f"{PIPELINE_LINE}\n[Open Deep Signal →]({app_url})"


def _story_block(story: dict) -> str:
    emoji = SEV_EMOJI.get(story["severity"], "🟡")
    return (
        f"{emoji} *{_s(story['headline'])}*\n"
        f"{_s(story['summary'])}\n"
        f"_Why it matters: {_s(story['why'])}_"
    )


def format_briefing(title: str, stories: list[dict], app_url: str) -> str:
    """Morning briefing / afternoon update — a list of important stories."""
    blocks = "\n\n".join(_story_block(s) for s in stories)
    return (
        f"*🛰 Deep Signal — {_s(title)}*\n_{_today()}_\n\n"
        f"{blocks}\n\n{_footer(app_url)}"
    )


def format_roundup(important: list[dict], mediums: list[dict], app_url: str) -> str:
    """Evening roundup — important stories plus a bundled list of MEDIUM ones."""
    head = (
        "\n\n".join(_story_block(s) for s in important)
        if important
        else "_A quiet day — nothing critical or high-priority broke._"
    )
    radar = ""
    if mediums:
        bullets = "\n".join(f"• {_s(m['headline'])}" for m in mediums[:6])
        radar = f"\n\n🟡 *Also on the radar*\n{bullets}"
    return (
        f"*🛰 Deep Signal — Evening Roundup*\n_{_today()}_\n\n"
        f"{head}{radar}\n\n{_footer(app_url)}"
    )


def format_breaking(items: list[dict], app_url: str) -> str:
    """Immediate alert from the every-2-hours monitor."""
    blocks = "\n\n".join(
        f"🔴 *{_s(it['headline'])}*\n{_s(it['summary'])}\n"
        f"_Why this is critical: {_s(it['why'])}_"
        for it in items
    )
    return f"🚨 *Deep Signal — Breaking*\n\n{blocks}\n\n[Open Deep Signal →]({app_url})"


def format_weekly(entries: list[dict], app_url: str) -> str:
    """Weekly roundup of the slower MEDIUM-severity stories."""
    blocks = "\n\n".join(
        f"🟡 *{_s(e['headline'])}*\n{_s(e['summary'])}" for e in entries[:12]
    )
    return (
        f"*🛰 Deep Signal — Weekly Roundup*\n"
        f"_The slower-moving stories from the past week_\n\n"
        f"{blocks}\n\n[Open Deep Signal →]({app_url})"
    )


def format_usage_warning(u: dict) -> str:
    """Cost-protection warning sent once when usage crosses the threshold."""
    return (
        f"⚠️ *Deep Signal — API usage warning*\n\n"
        f"Today's calls so far: Gemini {u['gemini']}/{u['maxGemini']}, "
        f"Groq {u['groq']}/{u['maxGroq']}.\n\n"
        f"If a daily cap is reached the system pauses and resumes tomorrow — "
        f"it stays free either way."
    )


def format_topic_reply(query: str, stories: list[dict], app_url: str) -> str:
    """Reply to a topic/question asked interactively via the Telegram bot."""
    if not stories:
        return (
            f"🛰 *Deep Signal*\n\n"
            f"I couldn't find anything solid on _{_s(query)}_ right now — "
            f"try rephrasing it."
        )
    blocks = "\n\n".join(_story_block(s) for s in stories)
    return f"*🛰 Deep Signal — {_s(query)}*\n\n{blocks}\n\n{_footer(app_url)}"


def format_news_list(title: str, stories: list[dict]) -> str:
    """Numbered list of today's stories — sent before the 'my take?' buttons."""
    lines = []
    for i, s in enumerate(stories, 1):
        emoji = SEV_EMOJI.get(s["severity"], "🟡")
        lines.append(
            f"{emoji} *{i}. {_s(s['headline'])}*\n"
            f"{_s(s['summary'])}\n"
            f"_{_s(s.get('domain', ''))} · why it matters: {_s(s['why'])}_"
        )
    body = "\n\n".join(lines)
    return (
        f"*🛰 Deep Signal — {_s(title)}*\n_{_today()}_\n\n{body}\n\n"
        f"👇 Want my take? Tap a story number below — or *All* for every story."
    )


def format_verdict_reply(
    topic: str, story: dict, verdict: dict, others: list[dict], app_url: str
) -> str:
    """The bot's full answer: the key story + Deep Signal's verdict."""
    sections: list[str] = []
    if verdict.get("analysis"):
        sections.append(f"🧠 *My read*\n{_s(verdict['analysis'])}")
    if verdict.get("opinion"):
        sections.append(f"💬 *My take*\n{_s(verdict['opinion'])}")
    if verdict.get("solution"):
        sections.append(f"🛠 *What should happen*\n{_s(verdict['solution'])}")
    if verdict.get("outcomes"):
        lines = "\n".join(
            f"• _{_s(o['horizon'])}_ — {_s(o['outcome'])}" for o in verdict["outcomes"]
        )
        sections.append(f"🔮 *Likely outcomes*\n{lines}")

    related = ""
    if others:
        related = "\n\n📌 *Related angles*\n" + "\n".join(
            f"• {_s(o['headline'])}" for o in others[:4]
        )

    emoji = SEV_EMOJI.get(story["severity"], "🟡")
    body = "\n\n".join(sections)
    return (
        f"*🛰 Deep Signal — {_s(topic)}*\n\n"
        f"{emoji} *{_s(story['headline'])}*\n{_s(story['summary'])}\n\n"
        f"{body}{related}\n\n{_footer(app_url)}"
    )

File: src/lib/briefing.ts
// Builds the Telegram briefing messages (Markdown). All dynamic text is run
// through sanitizeMarkdown so it can't break Telegram's parser.

import { sanitizeMarkdown } from "./telegram";
import type { LogEntry } from "./store";
import type { ScanItem, Severity, Story, Verdict } from "./types";
import type { UsageSnapshot } from "./usage";

const SEV_EMOJI: Record<Severity, string> = {
  CRITICAL: "🔴",
  HIGH: "🟠",
  MEDIUM: "🟡",
  LOW: "⚪",
};

const PIPELINE_LINE =
  "_Pipeline: Gemini 2.5 Flash + Flash-Lite · Groq Llama 4 Scout_";

function s(text: string): string {
  return sanitizeMarkdown(text);
}

function today(): string {
  return new Date().toLocaleDateString("en-GB", {
    weekday: "long",
    day: "numeric",
    month: "long",
  });
}

function footer(appUrl: string): string {
  return `${PIPELINE_LINE}\n[Open Deep Signal →](${appUrl})`;
}

function storyBlock(story: Story): string {
  return (
    `${SEV_EMOJI[story.severity]} *${s(story.headline)}*\n` +
    `${s(story.summary)}\n` +
    `_Why it matters: ${s(story.why)}_`
  );
}

/** Morning briefing / afternoon update — a list of important stories. */
export function formatBriefing(
  title: string,
  stories: Story[],
  appUrl: string,
): string {
  const blocks = stories.map(storyBlock).join("\n\n");
  return (
    `*🛰 Deep Signal — ${s(title)}*\n_${today()}_\n\n` +
    `${blocks}\n\n${footer(appUrl)}`
  );
}

/** Evening roundup — important stories plus a bundled list of slower MEDIUM ones. */
export function formatRoundup(
  important: Story[],
  mediums: Story[],
  appUrl: string,
): string {
  const head =
    important.length > 0
      ? important.map(storyBlock).join("\n\n")
      : "_A quiet day — nothing critical or high-priority broke._";

  let radar = "";
  if (mediums.length > 0) {
    const bullets = mediums
      .slice(0, 6)
      .map((m) => `• ${s(m.headline)}`)
      .join("\n");
    radar = `\n\n🟡 *Also on the radar*\n${bullets}`;
  }

  return (
    `*🛰 Deep Signal — Evening Roundup*\n_${today()}_\n\n` +
    `${head}${radar}\n\n${footer(appUrl)}`
  );
}

/** Immediate alert from the every-2-hours monitor. */
export function formatBreaking(items: ScanItem[], appUrl: string): string {
  const blocks = items
    .map(
      (it) =>
        `🔴 *${s(it.headline)}*\n${s(it.summary)}\n` +
        `_Why this is critical: ${s(it.why)}_`,
    )
    .join("\n\n");
  return (
    `🚨 *Deep Signal — Breaking*\n\n${blocks}\n\n` +
    `[Open Deep Signal →](${appUrl})`
  );
}

/** Weekly roundup of the slower MEDIUM-severity stories. */
export function formatWeekly(entries: LogEntry[], appUrl: string): string {
  const blocks = entries
    .slice(0, 12)
    .map((e) => `🟡 *${s(e.headline)}*\n${s(e.summary)}`)
    .join("\n\n");
  return (
    `*🛰 Deep Signal — Weekly Roundup*\n` +
    `_The slower-moving stories from the past week_\n\n` +
    `${blocks}\n\n[Open Deep Signal →](${appUrl})`
  );
}

/** Reply to a topic/question asked interactively via the Telegram bot. */
export function formatTopicReply(
  query: string,
  stories: Story[],
  appUrl: string,
): string {
  if (stories.length === 0) {
    return (
      `🛰 *Deep Signal*\n\n` +
      `I couldn't find anything solid on _${s(query)}_ right now — ` +
      `try rephrasing it.`
    );
  }
  const blocks = stories.map(storyBlock).join("\n\n");
  return `*🛰 Deep Signal — ${s(query)}*\n\n${blocks}\n\n${footer(appUrl)}`;
}

/**
 * The interactive bot's full answer to a topic: the key story, then Deep
 * Signal's own analysis, opinion, proposed solution and likely outcomes.
 */
export function formatVerdictReply(
  topic: string,
  story: Story,
  verdict: Verdict,
  others: Story[],
  appUrl: string,
): string {
  const sections: string[] = [];
  if (verdict.analysis) {
    sections.push(`🧠 *My read*\n${s(verdict.analysis)}`);
  }
  if (verdict.opinion) {
    sections.push(`💬 *My take*\n${s(verdict.opinion)}`);
  }
  if (verdict.solution) {
    sections.push(`🛠 *What should happen*\n${s(verdict.solution)}`);
  }
  if (verdict.outcomes.length > 0) {
    const lines = verdict.outcomes
      .map((o) => `• _${s(o.horizon)}_ — ${s(o.outcome)}`)
      .join("\n");
    sections.push(`🔮 *Likely outcomes*\n${lines}`);
  }

  const related =
    others.length > 0
      ? `\n\n📌 *Related angles*\n` +
        others
          .slice(0, 4)
          .map((o) => `• ${s(o.headline)}`)
          .join("\n")
      : "";

  return (
    `*🛰 Deep Signal — ${s(topic)}*\n\n` +
    `${SEV_EMOJI[story.severity]} *${s(story.headline)}*\n${s(story.summary)}\n\n` +
    `${sections.join("\n\n")}${related}\n\n${footer(appUrl)}`
  );
}

/** Cost-protection warning sent once when usage crosses the threshold. */
export function formatUsageWarning(u: UsageSnapshot): string {
  return (
    `⚠️ *Deep Signal — API usage warning*\n\n` +
    `Today's calls so far: Gemini ${u.gemini}/${u.maxGemini}, ` +
    `Groq ${u.groq}/${u.maxGroq}.\n\n` +
    `If a daily cap is reached the system pauses and resumes tomorrow — ` +
    `it stays free either way.`
  );
}
